### PR TITLE
Feature/local extensions

### PR DESCRIPTION
**Pull request by mwesterhof**
this is a proposed change to add a "_local_extensions" feature.
It will enable the use of an additional "_local_extensions" element in the cookiecutter.json file. This will have the exact same functionality as the "_extensions" element, except that it can be used to load jinja2 template extensions straight from the project template.

I'm not certain if this is the best approach to achieve this, but I'm happy to have some feedback, of course.

```json
{
    "directory_name": "directory_name",
    "module_name": "module_name",
    "_local_extensions": ["local_extensions.SomeExtension"]
}
```

### CODE REVIEW

**Comment by croesnick** [comment]
I totally missed your PR, @mwesterhof, when quickly scanning the open PRs today for exactly this feature. While not being finally decided whether I like an implicit (silent) loading of local extensions or your more explicit approach via dedicated keyword, I definitely like your context-manager approach for prepending `repo_dir` to `sys.path`. 🙂

**Comment by mwesterhof** [comment]
> I totally missed your PR, @mwesterhof, when quickly scanning the open PRs today for exactly this feature. While not being finally decided whether I like an implicit (silent) loading of local extensions or your more explicit approach via dedicated keyword, I definitely like your context-manager approach for prepending `repo_dir` to `sys.path`.

Thanks, it was a bit of a challenge to find the right way to implement this. I'd be very interested in this feature, though. When you decide on the way you want this to work, feel free to let me know. I'd be happy to make some changes

**Comment by insspb** [comment]
@mwesterhof @croesnick thank you for contribution. 
As you both are active, we can speed up this feature implementation. 
First it should be well documented from point how user should it use.

**Comment by mwesterhof** [comment]
@insspb @croesnick 
i added some docs for this feature to the "advanced" section. I haven't specified a version number yet, of course... should I leave that the way it is, or can i just replace the X.x with the next upcoming version?

**Comment by con-f-use** [comment]
I'm sorry, it seems I've implemented a very similar thing in #1340

**Comment by grahamalama** [comment]
Love this PR @mwesterhof! I was thinking about how to tackle the `modify config from a hook` problem that others were having, and basically came to the same idea that you did here. 

I had one idea that I thought might improve this implementation. I haven't fully thought it through, but I wanted to get your take:

As a `cookiecutter` user, I don't want to have to know / care about how to write a custom Jinja filter -- I just want to write a python function (or functions) that I can use in my `cookiecutter.json`.

What about instead of having users create a full-blown extension like:
```python
class FoobarExtension(Extension):
    def __init__(self, environment):
        super(FoobarExtension, self).__init__(environment)
         
        def foo():
             return "bar"
        
        environment.filters['foobar'] = foo
``` 

we had them create a module `filters.py` where they'd define a bunch of functions:
```python
def foo():
   return "bar"

def baz():
   return "qux"
```

and we could do something like:
```python
import inspect

import filters

functions = inspect.getmembers(filters, inspect.isfunction)
for function in functions:
    register_function_as_filter(function)
```

Thoughts on this?

**Comment by mwesterhof** [comment]
it's an interesting take, @grahamalama 
I'm generally in favor of that kind of simplification, but i have a feeling it might be a bit too restrictive and too "magical".

I'm no expert on jinja, but i think we're ignoring a lot of the power of template filters by simplifying it this much. I'd propose a compromise:

```python
@simple_filter
def foo():
    return 'stuff'
```

1. creating fully fledged jinja-based filters is still possible
2. creating a simple function based one is possible
3. registering the function based filters is explicit
4. it's possible to add other (helper) functions to the module without them erroneously being picked up as filters

it actually mimics some ideas in the django framework regarding template filters and template tags, and it's an approach that i'm a big fan of.

Any opinions? I'd be happy to implement this behavior in my branch

**Comment by con-f-use** [comment]
I still maintain, that just adding the extensions directory to the path and letting the user do whatever is the best, most simple and most generic solution.

**Comment by mwesterhof** [comment]
@con-f-use i would agree. importing the decorator as an additional helper from cookiecutter.utils is exactly what i had in mind. As far as separating the local extensions from the "regular" ones goes... i'm sure this is somewhat of a personal thing, but I'm personally a fan of having a clear distinction between "extensions bundled with the template" and "installed extensions". That's just my view on it, but i'd be happy to change that. I'm not entirely sure who makes that call, but please let me know

**Comment by mwesterhof** [comment]
@insspb @croesnick @con-f-use @grahamalama 
i just resolved a few conflicts, and I'm wondering if anyone has time to look at this MR. I'm more than happy to implement some changes, if necessary

**Comment by gasbasd** [comment]
any news about this?

File: tests/test_cli.py
Comment (grahamalama) [inline tests/test_cli.py:417]: Would it be good to include tests for non-happy path? I can think of two cases, but maybe there are more:
- I include a filter in my template, but didn't include the extension in the `_local_extensions` list
- I include a filter in my template and extension in the `_local_extensions` list, but haven't defined that extension (or it's somehow not available to be imported)
  - It looks like there's a [similar test](https://github.com/cookiecutter/cookiecutter/blob/master/tests/test_environment.py#L8:L15) for built-in Jinja extensions

File: docs/advanced/local_extensions.rst
Comment (grahamalama) [inline docs/advanced/local_extensions.rst:22]: If I only need one or two extensions and they're not that complicated, I can just define them directly in a module `local_extensions.py`, correct? Maybe instead of having the documentation suggest a more-complicated package setup, it could suggest simply that you have a module `local_extensions` available in the template root with extensions defined, with no reference to a folder or `main.py`. That would leave it up to the user to use a standalone module or a package structure depending on their needs.

File: tests/test_cli.py
Comment (mwesterhof) [reply tests/test_cli.py:417]: that's a good idea. i'll pick that up ASAP

File: docs/advanced/local_extensions.rst
Comment (mwesterhof) [reply docs/advanced/local_extensions.rst:22]: that seems fair. i'll make the change

**Comment by jonathan-s** [comment]
I was looking for exactly this. I'm glad that this in the works. My preference, would be to not introduce new naming such as the `_local_extensions` and instead use the already existing `_extensions`. From reading the code on how you are patching the python path that shouldn't be a problem.

File: docs/advanced/local_extensions.rst
Comment (jonathan-s) [reply docs/advanced/local_extensions.rst:22]: From what I gather reading the code you are adding the path of the local directory. So in effect you could create a folder that is named `extensions` that contains an `__init__.py` file. In that file you'll have the extensions and you'd refer to that in the cookiecutter.json file as `'extensions.MyExtension'` and it should all work.

**Comment by jonathan-s** [comment]
@mwesterhof 
```
it actually mimics some ideas in the django framework regarding template filters and template tags, and it's an approach that i'm a big fan of.

Any opinions? I'd be happy to implement this behavior in my branch
```

I like the `@simplefilter` approach as well.

**Comment by mwesterhof** [comment]
sorry guys, i've been quite sick recently and haven't gotten around to any work. I'll start working on this again though, so expect to see changes soon :muscle:

**Comment by jonathan-s** [comment]
Mention me and I'll be happy to give feedback on any code changes. I hope
your illness isn't around any longer.

**Comment by mwesterhof** [comment]
@grahamalama  @jonathan-s I updated my branch to include the simple_filter decorator, and rebased it. Can you guys take a look at it and let me know what you think about it? I'll try to check out your comment about a simple local_extensions.py asap, @grahamalama, but i think it should work just fine. 
also, the point about the naming scheme is a valid one, please weigh in on this. I'd be happy to change it so everything can be defined under '_extensions', without needing a separate '_local_extensions', i'd just like some more input on it first.

I finally have some time to work on this, and I hope we can do something cool with this MR soon. Please let me know what you guys think :muscle:

File: cookiecutter/utils.py
Comment (grahamalama) [inline cookiecutter/utils.py:111]: 😍

**Comment by jonathan-s** [comment]
@mwesterhof The simplefilter looks 👌. As mentioned by @con-f-use, I also think it'd be better to use `extensions` rather than `local_extensions`. It'll be one thing less to remember for the user of cookiecutter. I don't see any major advantages in having `extensions` and `local_extensions` separate. Are there any compelling reasons for that?

**Comment by mwesterhof** [comment]
@jonathan-s i think it's a valid question. I initially wanted to leave it separate, but looking at the endresult, i doubt there's a particularly good reason to keep it that way. It shouldn't be a problem to change it. I'll wait for the feedback, but I'll make the change locally in the mean time so I can verify that it works fine in all scenarios

**Comment by con-f-use** [comment]
My vote should be obvious... **no** `_local_extension` :grin:

**Comment by mwesterhof** [comment]
@con-f-use @grahamalama @jonathan-s i made this change, we're now using exclusively _extensions
please let me know if you agree, i can always revert it if necessary

**Comment by jonathan-s** [comment]
@insspb, it would be great if you could take a look at this. In my opinion this is ready to be merged.

**Comment by con-f-use** [comment]
Please merge this already, it's super-useful!

**Comment by mwesterhof** [comment]
@jonathan-s based on his activity, i think insspb might be unavailable. Do you know if anyone else might be able to take a look at this in the mean time?

**Comment by jonathan-s** [comment]
@ssbarnea Would it be possible for you to take a look at this?

**Comment by mwesterhof** [comment]
@jonathan-s is there a plan C? :sweat_smile:

**Comment by jonathan-s** [comment]
@mwesterhof Sometimes it can take a while in the open source world. It's frustrating I know. Hacktober is coming up in a month, so let's wait for that. It'll likely get some more activity going.

**Comment by ssbarnea** [comment]
I am afraid I cannot help with this PR, mainly because I personally do not see the use-case for it. If understand correctly this would mean that you will be able to use cc only from a specific location and the result would be dependent on your setup. To be this seems to make tool usage harder. Wouldn't a use of plugin mechanisms like pluggy allow it to be easier to extend. Installing an extension would be done using pip, same as cc itself.

**Comment by mwesterhof** [comment]
@ssbarnea to clarify the use case; primarily, i've come across quite a few cases where we would like to have a template extension that does something relatively simple, yet very specific for a template that we're maintaining. Essentially, we want more power in creating and maintaining a CC template. Of course, installing extensions through pip is, and should remain an option, but this PR allows for much, much more flexibility by bundling custom python code with the template

**Comment by jonathan-s** [comment]
@ssbarnea, As described in this issue -> https://github.com/pydanny/cookiecutter-djangopackage/issues/966 I would like to use a custom template filter `split` in the cookiecutter. Without this patch it's not possible. 

> If understand correctly this would mean that you will be able to use cc only from a specific location and the result would be dependent on your setup. 

I fail to see how this would happen.

**Comment by mwesterhof** [comment]
> @ssbarnea, As described in this issue -> [pydanny/cookiecutter-djangopackage#966](https://github.com/pydanny/cookiecutter-djangopackage/issues/966) I would like to use a custom template filter `split` in the cookiecutter. Without this patch it's not possible.
> 
> > If understand correctly this would mean that you will be able to use cc only from a specific location and the result would be dependent on your setup.
> 
> I fail to see how this would happen.

actually, i think this might be an important misunderstanding. to clarify how this works... a "local" extension is an extension that's *bundled* with the template itself. So no, it's absolutely not dependent on an individual setup. The extension would be both used by, and *provided by* the template.

**Comment by con-f-use** [comment]
Yes, it actually makes things a lot easier for the enduser. At the company I work for, we have a number of usecases for this and actually run on a fork, currently, that already has this.

**Comment by xvr-hlt** [comment]
Please merge when possible, this is something that would be really useful for our projects.

**Review by ssbarnea — APPROVED** [review]

**Comment by ssbarnea** [comment]
@insspb Can you please have another look at this and make a decision, eventually a merge if it looks ok to you. Thanks.

**Comment by indigoviolet** [comment]
@insspb will this be merged?

**Comment by brayanjuls** [comment]
Hey guys, do you have a workaround to enable this feature while we wait for @insspb to merge this feature ?

**Comment by jonathan-s** [comment]
The most sensible workaround would be to ask the user of the cookie cutter
template to install a forked cookiecutter, where the fork contains this
fix.

**Comment by mwesterhof** [comment]
@ssbarnea @jonathan-s apologies if I seem awfully pushy on this, but @insspb doesn't seem available for review, and hasn't been since he was first mentioned in august last year. Considering the fact that this regards a relatively small but welcome feature that doesn't impact any of the current features, I'm wondering if there's a way to short-track this a bit

**Comment by jonathan-s** [comment]
I don't have any power here. It'd be up to ssbarnea in that case.

**Comment by con-f-use** [comment]
Friendly reminder to @ssbarnea that we'd still very much like to have this.

File: cookiecutter/main.py
Comment (samj1912) [inline cookiecutter/main.py:7]: Any reason for adding this?

File: cookiecutter/main.py
Comment (samj1912) [inline cookiecutter/main.py:78]: This seems like it would add all the python files at the same level as the template directory to the import path. That might cause unexpected issues for people. Would it be better to limit the scope here to just the local extensions directory?

File: docs/advanced/local_extensions.rst
Comment (samj1912) [inline docs/advanced/local_extensions.rst:22]: Should this be updated to reflect the fact that we are adding the folder to sys path instead and the extension names would be derived from the modules/paths in the repo dir instead? The main.py may throw people off.

File: cookiecutter/main.py
Comment (mwesterhof) [reply cookiecutter/main.py:7]: good catch, it seems I added this because I needed it for an initial change. It's been over 1.5 years, and I guess I forgot about that. I'll remove it

File: cookiecutter/main.py
Comment (mwesterhof) [reply cookiecutter/main.py:78]: While it makes some sense, I'm not sure it would be the best way to go. 
The general idea with the current implementation is that template extension loading will essentially continue to work the same way it always has, except that they will also be loaded from local sources. While the provided example shows extensions in the "local_extensions" folder, I'm not suggesting that the extensions need to be contained in that specific folder, or even that extensions should all exist in the same folder.

I'm not convinced that it's necessary to limit the scope, but I'll make the change if I haven't changed your mind. I assume that adding another config keyword doesn't cause any problems? In that case I would suggest changing this:

```
    {
        "project_slug": "Foobar",
        "year": "{% now 'utc', '%Y' %}",
        "_extensions": ["local_extensions.FoobarExtension"]
    }
```
to 
```
    {
        "project_slug": "Foobar",
        "year": "{% now 'utc', '%Y' %}",
        "_local_extensions_folders": ["local_extensions"],
        "_extensions": ["FoobarExtension"]
    }
```
any thoughts?

File: docs/advanced/local_extensions.rst
Comment (mwesterhof) [reply docs/advanced/local_extensions.rst:22]: I'm honestly not sure what it is you're referring to. This is meant to be an example involving a python module as a "container" for extensions, and that's literally what is being explained here.

Maybe if we're indeed going with the "local extension folder config option" idea, then maybe this example could be simplified a bit, but the basic principle is kept as simple as possible while still being pythonic, and the changes in main.py are only to facilitate the extension loading and make it all possible

**Comment by jonathan-s** [comment]
Can we please get this in the 2.0 release?

**Comment by con-f-use** [comment]
@audreyfeldroy Pretty please? :grin:

**Comment by frankhuurman** [comment]
2 days ago I've opened up an issue requesting to use some sort of custom .py script or function while generating the cookiecutter project to set a value in the cookiecutter.json, I believe this feature is exactly what I'm looking for!
Is there any update on this?

This is the one feature needed to automate all the things with cookiecutter :stuck_out_tongue:

**Comment by con-f-use** [comment]
Yeah, please merge this asap, I'd prefer not to wait another 2.5 years.

**Comment by matrixise** [comment]
Hi, I hope the best for @insspb but unfortunately, there is no Github activities for his account since December 2021.

Could we review/merge this PR without him?

Thank you,

**Comment by mwesterhof** [comment]
@insspb indeed doesn't seem available long term. I'm sorry to once again have to bother @ssbarnea , but is there any way to proceed with this PR without his input?

**Comment by ssbarnea** [comment]
I was planning to merge it but I wanted to see the CI result first.

**Comment by mwesterhof** [comment]
@ssbarnea it seems that black linting was added somewhere along the line. I accounted for that, and it looks like everything passes now

**Comment by con-f-use** [comment]
I'm overjoyed :grin: Thanks ssabernea! :1st_place_medal:

**Comment by frankhuurman** [comment]
Awesome, can't wait to use it cause this will really make things a lot more extensive!

Same question as @con-f-use, is there a planned release date(with updated Docs as well?).

**Comment by connerxyz** [comment]
I'm eager make use of this feature too! Thanks.

File: docs/advanced/local_extensions.rst
Comment (alkatar21) [inline docs/advanced/local_extensions.rst:22]: ``main.py`` must be ``__init__.py``

**Comment by con-f-use** [comment]
Awesome, now we just have to wait for a new release. Any idea when that might be?

**Comment by grahamalama** [comment]
@con-f-use looks like it made it into [this release](https://github.com/cookiecutter/cookiecutter/releases/tag/2.1.0).

**Pull request merged**
